feat: Introduce sqlserver state store v2 (#5032)
dapr/docs: daprdocs/content/en/reference/components-reference/supported-state-stores/setup-sqlserver-v2.md

---
type: docs
title: "Microsoft SQL Server & Azure SQL"
linkTitle: "Microsoft SQL Server & Azure SQL"
description: Detailed information on the Microsoft SQL Server state store component
aliases:
  - "/operations/components/setup-state-store/supported-state-stores/setup-sqlserver-v2/"
---

{{% alert title="Note" color="primary" %}}

> This is the v2 of the Sql Server state store component, which contains full workflows support. New applications are encouraged to use v2.

Reviewer: ```suggestion
This is the v2 of the Sql Server state store component, which supports Dapr workflows. New applications are encouraged to use v2.
```

The state store does not have an index on the `ExpireDate` column, which means that each clean up operation must perform a full table scan. If you intend to write to the table with a large number of records that use TTLs, you should consider creating an index on the `ExpireDate` column. An index makes queries faster, but uses more storage space and slightly slows down writes.

```sql
CREATE CLUSTERED INDEX expiredate_idx ON state(ExpireDate ASC)
```

## Advanced

### Differences between v1 and v2

The Sqlserver state store v1 was introduced in Dapr 1.5. The [pre-existing v1]({{% ref setup-sqlserver.md %}}) remains available and is not deprecated.

In the v2 component, the table schema has been changed, with the goal of fully supporting workflows. Most notably, the value stored by Dapr is now of type _VARBINARY_, which allows workflows to store the data without errors.  
However, due to this change, the v2 component does not support the [Dapr state store query APIs]({{% ref howto-state-query-api.md %}}).

> Because of these changes, v1 and v2 components are not able to read or write data from the same table. At this stage, it's also impossible to migrate data between the two versions of the component.

Reviewer: ```suggestion
Because of these changes, v1 and v2 components are not able to read or write data from the same table. It is also not possible to migrate data between the two versions of the component.
```
